### PR TITLE
Add reach strategy implementation for custom reached logic

### PR DESCRIPTION
**Pull request by Jonatino**
Added a new interface for implementing custom reach checks. Has various use cases throughout RSPS implementations.

Please review and leave and suggestions below.

### CODE REVIEW

File: src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt [hunk 1]
```diff
@@ -23,17 +22,14 @@ private const val MAX_ALTERNATIVE_ROUTE_LOWEST_COST = 1000
...
-private const val RECTANGLE_STRATEGY = 2
-private const val NO_STRATEGY = 3
-
 /*
  * For optimization, we use this value to separate each section
  * where the list of route coordinates made a turn in any direction.
  */
 private val TURN_COORDS = RouteCoordinates(0)
 
+private val EMPTY_LIST = emptyList<RouteCoordinates>()
```
Comment (Tomm0017) [inline src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt:31]: `emptyList<T>()` returns `EmptyList` (an internal instance of an empty list). No need to create a new instance here

File: src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt [hunk 2]
```diff
@@ -713,12 +683,4 @@ public class SmartPathFinder(
         return this[index]
     }
```
Comment (Tomm0017) [inline src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt:685]: Empty newline

File: src/main/kotlin/org/rsmod/pathfinder/reach/ReachStrategy.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        destY: Int,
+        destWidth: Int,
+        destHeight: Int,
+        srcSize: Int,
+        rotation: Int,
+        shape: Int,
+        accessBitMask: Int,
+        searchMapSize: Int
+    ): Boolean
+
```
Comment (Tomm0017) [inline src/main/kotlin/org/rsmod/pathfinder/reach/ReachStrategy.kt:19]: Empty newline

File: src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt (hunk 1 above)
Comment (Jonatino) [reply src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt]: Good to know. Removed.

File: src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt (hunk 2 above)
Comment (Jonatino) [reply src/main/kotlin/org/rsmod/pathfinder/SmartPathFinder.kt]: Removed

File: src/main/kotlin/org/rsmod/pathfinder/reach/ReachStrategy.kt (hunk 3 above)
Comment (Jonatino) [reply src/main/kotlin/org/rsmod/pathfinder/reach/ReachStrategy.kt]: Removed

**Pull request merged**